File: DP/3904-Smallest-Stable-Index-II/main.go
package main

func firstStableIndex(nums []int, k int) int {
	l := len(nums)
	minCheck := make([]int, l)
	maxCheck := make([]int, l)
	minNum := nums[l-1]
	maxNum := nums[0]

	for i := 0; i < l; i++ {
		minNum = min(minNum, nums[l-1-i])
		minCheck[l-1-i] = minNum

		maxNum = max(maxNum, nums[i])
		maxCheck[i] = maxNum
	}

	for i := 0; i < l; i++ {
		if maxCheck[i]-minCheck[i] <= k {
			return i
		}
	}

	return -1
}

File: DP/3904-Smallest-Stable-Index-II/main.py
class Solution:
    def firstStableIndex(self, nums: list[int], k: int) -> int:
        l = len(nums)
        min_check = [0] * l
        min_num = nums[l - 1]
        max_check = [0] * l
        max_num = nums[0]

        for i in range(l):
            min_num = min(min_num, nums[l - 1 - i])
            min_check[l - 1 - i] = min_num

            max_num = max(max_num, nums[i])
            max_check[i] = max_num

        for i in range(l):
            if max_check[i] - min_check[i] <= k:
                return i

        return -1
